feat(client-presence): initial telemetry

jason-ha opened the pull request:
Add monitoring context `Presence`
Two events:
- `PresenceInitiated` logged when runtime instantiates `Presence` feature.
- `JoinResponse` logged when a client manages response to another client's Join message.

Add basic test infrastructure to check expected events and protocol handling.
- Add mock IEphemeralRuntime implementation with signal tracking
- Refactor SessionId assignment to allow testing to specify it (mocking underlying functions used generate it is not supported with ESM)
- Update/fix Presence test name hierarchy

jason-ha committed:
feat(client-presence): initial telemetry
Add monitoring context `Presence`
Two events:
- `PresenceInitiated` logged when runtime instantiates `Presence` feature.
- `JoinResponse` logged when a client manages response to another client's Join message.

markfields commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -256,9 +256,18 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+				this.presence.mc?.logger.sendTelemetryEvent({
+					eventName: "JoinResponse",
+					type: "broadcastAll",
+					// TODO: this is EUPI; how should it be tagged?
```
We don't currently tag EUPI.

markfields commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -256,9 +256,18 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
+					type: "broadcastAll",
+					// TODO: this is EUPI; how should it be tagged?
+					// If it can't be logged, then we can establish a correlation in the join request.
+					requestor,
+					role: "primary",
```
I'd nest these under a property called `details`. It makes it easier to discover this info when scanning through a session log, and you can easily parse the JSON to operate/aggregate on the data if needed.

For our 1P partner's telemetry DB, we've run out of columns, so these will need to be pulled out of a catch-all anyway when querying - better to do it proactively / predictably.

jason-ha committed:
feat(client-presence): update event data

markfields commented on packages/framework/presence/src/presenceDatastoreManager.ts:
```diff
@@ -256,9 +256,18 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
 			// If it isn't then, not really a problem; just won't be in provider or quorum list.
 			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
 			const clientId = this.runtime.clientId!;
+			const requestor = message.clientId;
```
Want to call it `requestorClientId` to make the semantics of it more clear?  I'm assuming it is some other client's clientId, same as this client has a clientId which we log all over.

jason-ha replied:
I did not because clientId is a bit confusing in this context. We do say clientId all over the place and mean client's curernt connection id. We explicitly differentiate within presence space because we mostly work with client session id (also a guid). I could name it requestorClientConnectionId, but that will be confusing and very verbose.
We don't actually care telemetry-wise that it is a client connection id - at least not for any of the metrics I am thinking about (also needs to be written out). What matters is that it is a unique id in response to a particular join (same id used by all respondents if there are multiple respondents).
So genericly named that happens to be client connection id is good for now. (Probably I should add some comments.)

markfields reviewed (approved):
LGTM! Couple bonus thoughts - 

- Any concerns about this log getting hit too much, in tight loop or on every signal or something?
- If you _want_ (and it's not always necessary/advisable) you could add unit tests to ensure the logging happens as you expect, using `MockLogger`.  At the very least I would step through these scenarios and make sure the logs are happening as expected once (even if you don't want to write/check in a test)

jason-ha commented:
- Neither of those can (well shouldn't) hit frequently. The first event should be no more than once per run of client. The second is in response to other clients joining session.
- There is a massive test debt with the experimental package. We have API compilation tests, but not execution tests. We'll definitely want to be checking telemetry when we can get there.

jason-ha committed:
refactor(client-presence): break out join response
and add notes about `requestor` use

jason-ha committed:
test(client-presence): telemetry logs
- add mock EphemeralRuntime with signal tracking
- refactor session id creation to allow testing to specify it explicitly (otherwise createSessionId will need to be placed in an alterable object before calling and testing will need access to stub (replace))
- update test name hierarchy

jason-ha reviewed (commented):
> - There is a massive test debt with the experimental package. We have API compilation tests, but not execution tests. We'll definitely want to be checking telemetry when we can get there.
 
Test debt is less massive now. Thanks to the nudge I got some infrastructure in place.
